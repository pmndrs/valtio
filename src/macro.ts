import type { NodePath } from '@babel/traverse'
import * as t from '@babel/types'
import { createMacro, MacroError } from 'babel-plugin-macros'
// @ts-ignore
import { addNamed } from '@babel/helper-module-imports'

const macro = ({ references }: any) => {
  references.useProxy?.forEach((path: NodePath) => {
    const hook = addNamed(path, 'useSnapshot', 'valtio')
    const proxy = (path.parentPath.get('arguments.0') as any).node
    if (!t.isIdentifier(proxy)) throw new MacroError('no proxy object')
    const snap = t.identifier(`valtio_macro_snap_${proxy.name}`)
    path.parentPath.parentPath.replaceWith(
      t.variableDeclaration('const', [
        t.variableDeclarator(snap, t.callExpression(hook, [proxy])),
      ])
    )
    let inFunction = 0
    path.parentPath.getFunctionParent()?.traverse({
      Identifier(p) {
        if (
          inFunction === 0 && // in render
          p.node !== proxy &&
          p.node.name === proxy.name
        ) {
          p.node.name = snap.name
        }
      },
      Function: {
        enter() {
          ++inFunction
        },
        exit() {
          --inFunction
        },
      },
    })
  })
}

export declare function useProxy<T extends object>(proxyObject: T): void

export default createMacro(macro, { configName: 'valtio' })
