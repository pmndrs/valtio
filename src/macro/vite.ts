// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import { addNamed } from '@babel/helper-module-imports'
import * as t from '@babel/types'
import {
  createMacroPlugin,
  defineMacro,
  defineMacroProvider,
} from 'aslemammad-vite-plugin-macro'
import { MacroError } from 'babel-plugin-macros'

export const valtioMacro = defineMacro(`useProxy`)
  .withSignature(`<T extends object>(proxyObject: T): void`)
  .withHandler((ctx) => {
    const { path, args } = ctx
    const hook = addNamed(path, 'useSnapshot', 'valtio')
    const proxy = args[0]?.node

    if (!t.isIdentifier(proxy)) throw new MacroError('no proxy object')

    const snap = t.identifier(`valtio_macro_snap_${proxy.name}`)
    path.parentPath?.replaceWith(
      t.variableDeclaration('const', [
        t.variableDeclarator(snap, t.callExpression(hook, [proxy])),
      ])
    )

    let inFunction = 0
    path.parentPath?.getFunctionParent()?.traverse({
      Identifier(p) {
        if (
          inFunction === 0 && // in render
          p.node !== proxy &&
          p.node.name === proxy.name
        ) {
          p.node.name = snap.name
        }
      },
      Function: {
        enter() {
          ++inFunction
        },
        exit() {
          --inFunction
        },
      },
    })
  })

export function provideValtioMacro() {
  return defineMacroProvider({
    id: 'valtio/macro',
    exports: {
      'valtio/macro': {
        macros: [valtioMacro],
      },
    },
  })
}

const macroPlugin = createMacroPlugin({}).use(provideValtioMacro())

export default macroPlugin
