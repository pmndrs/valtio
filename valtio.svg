<svg width="1283" height="1869" viewBox="0 0 1283 1869" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect width="1283" height="1869" fill="white"/>
<path d="M13.2 24.9999H135.3L214.5 622.3H216.7L295.9 24.9999H407L290.4 795H129.8L13.2 24.9999ZM483.467 24.9999H647.367L772.767 795H651.767L629.767 642.1V644.3H492.267L470.267 795H358.067L483.467 24.9999ZM615.467 539.8L561.567 159.2H559.367L506.567 539.8H615.467ZM787.471 24.9999H908.471V685H1107.57V795H787.471V24.9999Z" fill="black"/>
<path d="M13.2 24.9999V23.9999H12.0372L12.2113 25.1497L13.2 24.9999ZM135.3 24.9999L136.291 24.8685L136.176 23.9999H135.3V24.9999ZM214.5 622.3L213.509 622.431L213.624 623.3H214.5V622.3ZM216.7 622.3V623.3H217.576L217.691 622.431L216.7 622.3ZM295.9 24.9999V23.9999H295.024L294.909 24.8685L295.9 24.9999ZM407 24.9999L407.989 25.1497L408.163 23.9999H407V24.9999ZM290.4 795V796H291.26L291.389 795.15L290.4 795ZM129.8 795L128.811 795.15L128.94 796H129.8V795ZM13.2 25.9999H135.3V23.9999H13.2V25.9999ZM134.309 25.1314L213.509 622.431L215.491 622.169L136.291 24.8685L134.309 25.1314ZM214.5 623.3H216.7V621.3H214.5V623.3ZM217.691 622.431L296.891 25.1314L294.909 24.8685L215.709 622.169L217.691 622.431ZM295.9 25.9999H407V23.9999H295.9V25.9999ZM406.011 24.8502L289.411 794.85L291.389 795.15L407.989 25.1497L406.011 24.8502ZM290.4 794H129.8V796H290.4V794ZM130.789 794.85L14.1887 24.8502L12.2113 25.1497L128.811 795.15L130.789 794.85ZM483.467 24.9999V23.9999H482.617L482.48 24.8392L483.467 24.9999ZM647.367 24.9999L648.354 24.8392L648.218 23.9999H647.367V24.9999ZM772.767 795V796H773.943L773.754 794.839L772.767 795ZM651.767 795L650.777 795.142L650.901 796H651.767V795ZM629.767 642.1L630.757 641.958L628.767 642.1H629.767ZM629.767 644.3V645.3H630.767V644.3H629.767ZM492.267 644.3V643.3H491.403L491.278 644.156L492.267 644.3ZM470.267 795V796H471.132L471.257 795.144L470.267 795ZM358.067 795L357.08 794.839L356.891 796H358.067V795ZM615.467 539.8V540.8H616.619L616.457 539.66L615.467 539.8ZM561.567 159.2L562.557 159.06L562.436 158.2H561.567V159.2ZM559.367 159.2V158.2H558.496L558.377 159.063L559.367 159.2ZM506.567 539.8L505.577 539.663L505.419 540.8H506.567V539.8ZM483.467 25.9999H647.367V23.9999H483.467V25.9999ZM646.38 25.1607L771.78 795.161L773.754 794.839L648.354 24.8392L646.38 25.1607ZM772.767 794H651.767V796H772.767V794ZM652.757 794.858L630.757 641.958L628.777 642.242L650.777 795.142L652.757 794.858ZM628.767 642.1V644.3H630.767V642.1H628.767ZM629.767 643.3H492.267V645.3H629.767V643.3ZM491.278 644.156L469.278 794.856L471.257 795.144L493.257 644.444L491.278 644.156ZM470.267 794H358.067V796H470.267V794ZM359.054 795.161L484.454 25.1607L482.48 24.8392L357.08 794.839L359.054 795.161ZM616.457 539.66L562.557 159.06L560.577 159.34L614.477 539.94L616.457 539.66ZM561.567 158.2H559.367V160.2H561.567V158.2ZM558.377 159.063L505.577 539.663L507.558 539.937L560.358 159.337L558.377 159.063ZM506.567 540.8H615.467V538.8H506.567V540.8ZM787.471 24.9999V23.9999H786.471V24.9999H787.471ZM908.471 24.9999H909.471V23.9999H908.471V24.9999ZM908.471 685H907.471V686H908.471V685ZM1107.57 685H1108.57V684H1107.57V685ZM1107.57 795V796H1108.57V795H1107.57ZM787.471 795H786.471V796H787.471V795ZM787.471 25.9999H908.471V23.9999H787.471V25.9999ZM907.471 24.9999V685H909.471V24.9999H907.471ZM908.471 686H1107.57V684H908.471V686ZM1106.57 685V795H1108.57V685H1106.57ZM1107.57 794H787.471V796H1107.57V794ZM788.471 795V24.9999H786.471V795H788.471Z" fill="black" fill-opacity="0.3"/>
<path d="M457.966 927H331.466V817H705.466V927H578.966V1587H457.966V927ZM720.049 817H841.049V1587H720.049V817ZM1062.57 1598C1003.17 1598 957.704 1581.13 926.17 1547.4C894.637 1513.67 878.87 1466 878.87 1404.4V999.6C878.87 938 894.637 890.333 926.17 856.6C957.704 822.867 1003.17 806 1062.57 806C1121.97 806 1167.44 822.867 1198.97 856.6C1230.5 890.333 1246.27 938 1246.27 999.6V1404.4C1246.27 1466 1230.5 1513.67 1198.97 1547.4C1167.44 1581.13 1121.97 1598 1062.57 1598ZM1062.57 1488C1104.37 1488 1125.27 1462.7 1125.27 1412.1V991.9C1125.27 941.3 1104.37 916 1062.57 916C1020.77 916 999.87 941.3 999.87 991.9V1412.1C999.87 1462.7 1020.77 1488 1062.57 1488Z" fill="black"/>
</svg>
