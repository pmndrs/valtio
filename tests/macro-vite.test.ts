import { EnvContext } from '@typed-macro/core'
import { createTransformer } from '@typed-macro/runtime'
import { valtioMacro } from 'valtio/macro/vite'

const env: EnvContext = {
  host: 'test',
  packageManager: 'test',
  projectPath: [''],
  dev: true,
  ssr: false,
}

it('basic', async () => {
  const transform = createTransformer({})
  transform.appendMacros('valtio/macro', [valtioMacro])

  expect(
    transform.transform(
      `
import { useProxy } from 'valtio/macro'

const Component = () => {
  useProxy(state)
  return (
    <div>
      {state.count}
      <button onClick={() => ++state.count}>inc</button>
    </div>
  )
}
`,
      'test.ts',
      env
    )
  ).toMatchSnapshot()
})

it('complex', async () => {
  const transform = createTransformer({})
  transform.appendMacros('valtio/macro', [valtioMacro])

  expect(
    transform.transform(
      `
import { useProxy } from 'valtio/macro'

const Component = () => {
  useProxy(state)
  return (
    <div>
      <button onClick={() => {
        ;(() => ++state.count)()
        ++state.count
      }}>inc</button>
      {state.count}
    </div>
  )
}
`,
      'test.ts',
      env
    )
  ).toMatchSnapshot()
})
