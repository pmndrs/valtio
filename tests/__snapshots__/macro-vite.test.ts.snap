// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic 1`] = `
"import { useSnapshot as _useSnapshot } from "valtio";


const Component = () => {const valtio_macro_snap_state = _useSnapshot(
  state);
  return (
    <div>
      {valtio_macro_snap_state.count}
      <button onClick={() => ++state.count}>inc</button>
    </div>);

};"
`;

exports[`complex 1`] = `
"import { useSnapshot as _useSnapshot } from "valtio";


const Component = () => {const valtio_macro_snap_state = _useSnapshot(
  state);
  return (
    <div>
      <button onClick={() => {
        ;(() => ++state.count)();
        ++state.count;
      }}>inc</button>
      {valtio_macro_snap_state.count}
    </div>);

};"
`;
